gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonExceptionCallable.java

```diff
@@ -0,0 +1,127 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.client.http.HttpResponseException;
+import com.google.api.core.AbstractApiFuture;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.ApiExceptionFactory;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import java.util.Set;
+import java.util.concurrent.CancellationException;
+
+/**
+ * Transforms all {@code Throwable}s thrown during a call into an instance of {@link ApiException}.
+ *
+ * <p>Package-private for internal use.
+ */
+class HttpJsonExceptionCallable<RequestT, ResponseT> extends UnaryCallable<RequestT, ResponseT> {
+  private final UnaryCallable<RequestT, ResponseT> callable;
+  private final ImmutableSet<StatusCode.Code> retryableCodes;
+
+  HttpJsonExceptionCallable(
+      UnaryCallable<RequestT, ResponseT> callable, Set<StatusCode.Code> retryableCodes) {
+    this.callable = Preconditions.checkNotNull(callable);
+    this.retryableCodes = ImmutableSet.copyOf(retryableCodes);
+  }
+
+  @Override
+  public ApiFuture<ResponseT> futureCall(RequestT request, ApiCallContext inputContext) {
+    HttpJsonCallContext context =
+        HttpJsonCallContext.getAsHttpJsonCallContextWithDefault(inputContext);
+    ApiFuture<ResponseT> innerCallFuture = callable.futureCall(request, context);
+    ExceptionTransformingFuture transformingFuture =
+        new ExceptionTransformingFuture(innerCallFuture);
+    ApiFutures.addCallback(innerCallFuture, transformingFuture);
+    return transformingFuture;
+  }
+
+  private class ExceptionTransformingFuture extends AbstractApiFuture<ResponseT>
+      implements ApiFutureCallback<ResponseT> {
+    private ApiFuture<ResponseT> innerCallFuture;
+    private volatile boolean cancelled = false;
+
+    public ExceptionTransformingFuture(ApiFuture<ResponseT> innerCallFuture) {
+      this.innerCallFuture = innerCallFuture;
+    }
+
+    @Override
+    protected void interruptTask() {
+      cancelled = true;
+      innerCallFuture.cancel(true);
+    }
+
+    @Override
+    public void onSuccess(ResponseT r) {
+      super.set(r);
+    }
+
+    @Override
+    public void onFailure(Throwable throwable) {
+      StatusCode.Code statusCode = StatusCode.Code.UNKNOWN;
+      boolean canRetry = false;
```

Comment:
Please check the corresponding comment for `GrpcExceptionCallable#onFailure()` (same concern here).